### PR TITLE
PDS-479: CloudSQL query support

### PR DESCRIPTION
**Pull request by smitalimaye**
…d password

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -14,19 +14,20 @@ RUN pyb -q install
 #Upload the built python packge to pypi
 ARG PYPI_USERNAME
 ARG PYPI_PASSWORD
-RUN set -e && if [ "$PYPI_USERNAME" != "" ]; then twine upload /home/cloudsdk/app/target/dist/stanford-wdl-kit-1.3.0/dist/* -u $PYPI_USERNAME -p $PYPI_PASSWORD ; fi;
+RUN set -e && if [ "$PYPI_USERNAME" != "" ]; then twine upload /home/cloudsdk/app/target/dist/stanford-wdl-kit-1.3.0-slimaye9/dist/* -u $PYPI_USERNAME -p $PYPI_PASSWORD ; fi;
```
Comment (smitalimaye) [inline Dockerfile]: This needs to be fixed

File: Dockerfile [hunk 2]
```diff
@@ -14,19 +14,20 @@ RUN pyb -q install
...
 
 # Final container, copies package from above
 FROM google/cloud-sdk:395.0.0
 WORKDIR /home/cloudsdk/app
 RUN chown cloudsdk:cloudsdk /home/cloudsdk/app
 RUN apt-get install -y zip
 USER cloudsdk
-COPY --chown=cloudsdk:cloudsdk --from=build /home/cloudsdk/app/target/dist/stanford-wdl-kit-1.3.0/dist/stanford-wdl-kit-1.3.0.tar.gz /home/cloudsdk/app
+COPY --chown=cloudsdk:cloudsdk --from=build /home/cloudsdk/app/target/dist/stanford-wdl-kit-1.3.0-slimaye9/dist/stanford-wdl-kit-1.3.0-slimaye9.tar.gz /home/cloudsdk/app
```
Comment (smitalimaye) [inline Dockerfile]: This needs to be fixed

File: Makefile [hunk 3]
```diff
@@ -1,5 +1,5 @@
 TARGET = install
-VERSION=1.3.0
+VERSION=1.3.0-slimaye9
```
Comment (smitalimaye) [inline Makefile]: This needs to be fixed

File: build.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 name = "stanford-wdl-kit"
 default_task = "publish"
 summary = "A WDL toolkit with a focus on ETL and Cloud integration"
-version = "1.3.0"
+version = "1.3.0-slimaye9"
```
Comment (smitalimaye) [inline build.py]: This needs to be fixed

File: Makefile [hunk 5]
```diff
@@ -9,10 +9,13 @@ docker:
 	docker build -t wdl-kit:$(VERSION) .
 
 install: 
-	pip3 install pybuilder==0.13.5
+	pip3 install pybuilder==0.13.7
```
Comment (smitalimaye) [inline Makefile]: change to >=

File: src/main/python/utils/cloudsql.py [hunk 6]
```diff
@@ -21,6 +21,66 @@
...
+    # kind: str
+    name: str
+    instance: str
+    user: str
+    password: str
+    #email: str
+    query: str
+    connector: Connector
+
+    # initialize Connector object
```
Comment (smitalimaye) [inline src/main/python/utils/cloudsql.py]: Take this out

File: src/main/wdl/cloudsql.wdl [hunk 7]
```diff
@@ -134,3 +134,33 @@ task DeleteDatabase {
...
+task CsqlQuery { 
+    parameter_meta {
+        queryConfig: { description: "CsqlConfig object containing database connection details and query" }
+    }
+    input {
+        CsqlConfig queryConfig
+
+        Int cpu = 1
+        String memory = "128 MB"
+        String dockerImage = "wdl-kit:1.3.0-slimaye9"
```
Comment (smitalimaye) [inline src/main/wdl/cloudsql.wdl]: Take it out

File: src/main/wdl/cloudsql.wdl [hunk 8]
```diff
@@ -134,3 +134,33 @@ task DeleteDatabase {
...
+    }
+
+
+    command {
+        csql query ~{write_json(queryConfig)}
+    }
+
+    output {
+        File output_postgres = stdout()
```
Comment (smitalimaye) [inline src/main/wdl/cloudsql.wdl]: stdout and stderr

File: src/main/wdl/structs.wdl [hunk 9]
```diff
@@ -479,4 +479,12 @@ struct Database {
 struct SqlserverDatabaseDetails {
   String? compatibilityLevel
   String? recoveryModel
+}
+
+struct CsqlConfig {
```
Comment (smitalimaye) [inline src/main/wdl/structs.wdl:484]: Test it with DatabaseInstance

File: Makefile [hunk 10]
```diff
@@ -9,10 +9,13 @@ docker:
 	docker build -t wdl-kit:$(VERSION) .
 
 install: 
-	pip3 install pybuilder==0.13.5
+	pip3 install pybuilder>=0.13.7
```
Comment (smitalimaye) [inline Makefile:12]: Is this right?

File: src/main/python/utils/cloudsql.py [hunk 11]
```diff
@@ -21,6 +21,56 @@
...
+            creator=self.getconn,
+        )
+        
+        with engine.connect(self) as db_conn:
+
+            # query database
+            result = db_conn.execute(sqlalchemy.text(self.query))
+
+            # close connection
+            db_conn.close()
```
Comment (smitalimaye) [inline src/main/python/utils/cloudsql.py:73]: This might be redundant

File: src/main/python/utils/cloudsql.py [hunk 12]
```diff
@@ -21,6 +21,56 @@
...
+                enable_iam_auth=True if self.user.endswith('.iam') else False
+            )
+                
+        return conn
+
+    def queryDb(self):
+        
+        engine = sqlalchemy.create_engine(
+            "postgresql+pg8000://",
+            creator=self.getconn,
```
Comment (smitalimaye) [inline src/main/python/utils/cloudsql.py]: Take the comma out

File: src/main/wdl/cloudsql.wdl [hunk 13]
```diff
@@ -134,3 +134,38 @@ task DeleteDatabase {
...
+    output {
+        File stdout = stdout()
+        File stderr = stderr()
+    }
+
+    runtime {
+        docker: dockerImage
+        cpu: cpu
+        memory: memory
+        zones: queryConfig.region
```
Comment (smitalimaye) [inline src/main/wdl/cloudsql.wdl:168]: Setting the zone here, doesn't work. It prints a warning while running miniwdl saying -'ignored runtime settings :: keys: ["zones"]'
But if you create your instance in west1 zone, the docker container always seem to be created in west1 zone so far when I tested in dev project. 
cc @jmesterh Can we rely on the zone for docker container always being west1 and created db instances always in west1 zone?

Comment (jmesterh) [reply src/main/wdl/cloudsql.wdl:168]: this is only applicable if use are using cromwell server to run the WDL + the google backend selected in cromwell , so no. But its fine to assume that most things will be in the same zone.

**Review by jmesterh — APPROVED** [review]

**Pull request merged**